rmw_fastrtps_shared_cpp/src/participant.cpp

```diff
@@ -298,11 +342,70 @@ rmw_fastrtps_shared_cpp::destroy_participant(CustomParticipantInfo * participant
     RMW_SET_ERROR_MSG("participant_info is null");
     return RMW_RET_ERROR;
   }
-  Domain::removeParticipant(participant_info->participant);
-  delete participant_info->listener;
-  participant_info->listener = nullptr;
+
+  RMW_CHECK_ARGUMENT_FOR_NULL(participant_info->participant_, RMW_RET_ERROR);
+
+  // Make the participant stop listening to discovery
+  participant_info->participant_->set_listener(nullptr);
+
+  ReturnCode_t ret = ReturnCode_t::RETCODE_OK;
+
+  // Collect topics that should be deleted
+  std::vector<const eprosima::fastdds::dds::TopicDescription *> topics_to_remove;
+
+  // Remove datawriters and publisher from participant
+  if (participant_info->publisher_) {
+    std::vector<eprosima::fastdds::dds::DataWriter *> writers;
+    participant_info->publisher_->get_datawriters(writers);
+    for (auto writer : writers) {
+      topics_to_remove.push_back(writer->get_topic());
+      participant_info->publisher_->delete_datawriter(writer);
+    }
+    ret = participant_info->participant_->delete_publisher(participant_info->publisher_);
+    if (ret != ReturnCode_t::RETCODE_OK) {
+      RMW_SET_ERROR_MSG("Fail in delete dds publisher from participant");
+      return rmw_fastrtps_shared_cpp::cast_error_dds_to_rmw(ret);
```

Comment:
@MiguelCompany do not partially finalize. You may fail early on an invalid argument with no side effects, but otherwise finalization must be carried out to completion. See [`rmw_destroy_node()` API](https://github.com/ros2/rmw/blob/c71e97a670717cbf755627723a6d7ef5d756c4f4/rmw/include/rmw/rmw.h#L174-L183) contract. 

The same applies to every other finalization/destruction function that follows.